# GO富集分析
1. 提取差异基因
```r
# padj 小于 0.05 并且 Log2FC 大于 1（2倍） 或者小于 -1（1/2倍）
diff_gene <- subset(result_order, padj < 0.05 & abs(log2FoldChange) > 1)
# 查看数据框的大小
dim(diff_gene)
```
2. 绘制peak在基因结构上的分布情况
```r
# Load libraries
BiocManager::install("ChIPseeker")
BiocManager::install("GenomicFeatures")
BiocManager::install("TxDb.Mmusculus.UCSC.mm10.knownGene", force = TRUE)
BiocManager::install("org.Mm.eg.db", force = TRUE)
BiocManager::install("clusterProfiler", force = TRUE)

library(ChIPseeker)
library(GenomicFeatures)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(clusterProfiler)


# ① 导入bed文件
getwd()
# [1] "D:/brain/brain/R_analyse"

#  21531 HIPP_pool_merge.bed
#  45265 cortex_pool_merge.bed
#  58240 PFC_pool_merge.bed

HIPP_peak <- readPeakFile("D:/brain/brain/IDR_final/mouse/HIPP_pool_merge.bed",sep ="") 
# peak在染色体上的分布
covplot(HIPP_peak)

# peak 在TSS位点附件的分布
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
  promoter <- getPromoters(TxDb=txdb, upstream=1000, downstream=1000)
  tagMatrix <- getTagMatrix(HIPP_peak, windows=promoter)
tagHeatmap(tagMatrix, xlim=c(-1000, 1000), color="red")
plotAvgProf(
  tagMatrix,
  xlim=c(-1000, 1000),
  xlab="Genomic Region (5'->3')",
  ylab = "Peak Frequency")


# ② peak关联基因注释  
# 给出了关联的基因以及对应的基因组区域的类别，根据这个结果，可以提取关联基因进行下游的功能富集分析，比如提取geneid这一列，用clusterProfiler进行GO/KEGG等功能富集分析。  

HIPP_peakAnnolist <- annotatePeak(
    HIPP_peak,
    tssRegion = c(-1000, 1000),
    TxDb = txdb,
    annoDb = "org.Mm.eg.db")
HIPP_peakAnnolist

# Annotated peaks generated by ChIPseeker
# 21531/21531  peaks were annotated
# Genomic Annotation Summary:
#              Feature   Frequency
# 9           Promoter 49.52394222
# 4             5' UTR  0.34833496
# 3             3' UTR  1.84849752
# 1           1st Exon  1.79740839
# 7         Other Exon  3.09785890
# 2         1st Intron  6.79949840
# 8       Other Intron 14.78798012
# 6 Downstream (<=300)  0.06966699
# 5  Distal Intergenic 21.72681250

write.table(
    as.data.frame(HIPP_peakAnnolist),
    "HIPP_allpeak.annotation.tsv",
    sep="\t",
    row.names = F,
    quote = F)   

#可视化
plotAnnoPie(HIPP_peakAnnolist)
```
3. GO分析
```r
# ③ 基因ID转化  
# 提取差异基因
HIPP_peakAnno <- as.data.frame(HIPP_peakAnnolist)
# 转换函数
ensembl_id_transform <- function(ENSEMBL_ID){
    # geneID是输入的基因ID，fromType是输入的ID类型，toType是输出的ID类型，OrgDb注释的db文件，drop表示是否剔除NA数据
    a = bitr(ENSEMBL_ID, fromType="ENSEMBL", toType=c("SYMBOL","ENTREZID"), OrgDb="org.Mm.eg.db")
    return(a)
}
HIPP_ensembl_id_transform <- ensembl_id_transform(HIPP_peakAnno$ENSEMBL)
# 0.01% of input gene IDs are fail to map...  # 12498

# 写入文件
write.csv(HIPP_ensembl_id_transform), file="HIPP_allpeak_geneID.tsv", quote = F)
dim(HIPP_ensembl_id_transform)


# 使用ClusterProfiler包进行转化有一部分部分没有映射到，换biomaRt包试一下
BiocManager::install("biomaRt", force = TRUE)
library(biomaRt)
mart <- useDataset( "mmusculus_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL"))
 
HIPP_biomart_ensembl_id_transform <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id", "description"), filters = 'ensembl_gene_id', values = HIPP_peakAnno$ENSEMBL, mart = mart) # 转化了12496
write.csv(biomart_ensembl_id_transform, file="HIPP_allpeak_biomart_geneID.tsv", quote = F)


# ④ 功能富集分析 
# 选择clusterprofiler找到的基因进行分析   
# Run GO enrichment analysis 
HIPP_BP <- enrichGO(
        gene = HIPP_ensembl_id_transform$ENTREZID, 
        keyType = "ENTREZID",
        OrgDb = org.Mm.eg.db, 
        ont = "BP", 
        pAdjustMethod = "BH", 
        qvalueCutoff = 0.05, 
        readable = TRUE)
# bar visualization
barplot(HIPP_BP, showCategory=40, font.size = 6, title = paste("The GO BP enrichment analysis", sep = ""))
pdf(file="GO_BP.pdf")
barplot(HIPP_BP, showCategory=40, font.size = 6, title = paste("The GO BP enrichment analysis", sep = ""))
dev.off()


# KEGG
# Multiple samples KEGG analysis
HIPP_kegg <- enrichKEGG(gene =
        HIPP_ensembl_id_transform$ENTREZID,
        organism = 'mmu',
        pvalueCutoff = 0.05,
        pAdjustMethod = "BH")
barplot(HIPP_kegg, showCategory = 20, title = "KEGG Pathway Enrichment Analysis")

# 循环画图
for(i in c("MF", "BP", "CC")){
    ego <- enrichGO(gene       = rat_symbols$ensembl_gene_id,
                    OrgDb      = org.Rn.eg.db,
                    keyType    = 'ENSEMBL',
                    ont        = i,
                    pAdjustMethod = "BH",
                    pvalueCutoff = 0.01,
                    qvalueCutoff = 0.05)
   
dotplot(ego, showCategory = 30, title = paste("The GO ", i, " enrichment analysis", sep = ""))
}


barplot(ego, showCategory=20,title="GO_i")
    #条状图，按p从小到大排，绘制前20个Term

### 有向无环图goplot : Gene Ontology is organized as a directed acyclic graph.
gop <- goplot(ego, showCategory = 10)
ggsave(gop, filename = "goplot.pdf", width=10, height=10) 

## Tree plot  树图
BiocManager::install("ggnewscale")
library(clusterProfiler)
library(org.Rn.eg.db)
library(enrichplot)
library(GOSemSim)
library(DOSE)
library(ggnewscale)

pt <- pairwise_termsim(ego)
treep <- treeplot(pt,
                  showCategory = 30)
ggsave(treep, filename = 'treeplot.pdf', width=18, height=10)
```
